lazy_dataset/core.py

```diff
@@ -2313,6 +2397,95 @@ def __iter__(self):
             print(f'Dropped {dropped_count} examples')
 
 
+class CacheDataset(Dataset):
+    def __init__(self, dataset: Dataset, keep_mem_free=None,
+                 immutable_warranty: str = 'pickle') -> None:
+        super().__init__()
+        assert dataset.indexable, (
+            'Cache Dataset only works if dataset is indexable!'
+        )
+        self.dataset = dataset
+        self.cache = {}
+
+        self._keep_mem_free = self._get_memory_size(keep_mem_free)
+        self.immutable_warranty = immutable_warranty
+
+        if immutable_warranty == 'pickle':
+            self._serialize = pickle.dumps
+            self._deserialize = pickle.loads
+        elif immutable_warranty == 'copy':
+            self._serialize = lambda x: x
+            self._deserialize = deepcopy
+        else:
+            raise ValueError(immutable_warranty)
+
+    @property
+    def indexable(self) -> bool:
+        return True
+
+    def keys(self) -> list:
+        return self.dataset.keys()
+
+    @staticmethod
+    def _get_memory_size(keep_mem_free):
+        if keep_mem_free is None:
+            return None
+
+        assert isinstance(keep_mem_free, (list, tuple)) \
+               and len(keep_mem_free) == 2, (
+            f'keep_mem_free has to be a tuple of value and unit, and not '
+            f'{keep_mem_free}'
+        )
+
+        value, unit = keep_mem_free
```

Comment:
Currently, I am unsure, if `(5, 'GB')` or `'5 GB'` is better.
I think ESPnet uses `humanfriendly.parse_size` to parse such a str (At least it has humanfriendly as dependency).

Reply:
I was also not sure about the format of this. The first one makes it (slightly) easier to insert computed values, but the second one is more intuitive. I think I prefer the second one (str only)

Reply:
I think we do not need here computed values. So values like `10%` or `5GB` may be more intuitive.

Reply:
Done, now using humanfriendly